### PR TITLE
Metrics diff

### PR DESCRIPTION
**Pull request by malte-b**
*Description of changes:*
We implemented that it is possible to to compare multiple metrics repositories by combining them into one where the values of each repository are identified with a unique result key. With the resulting repository it is rather easy to display all the differences in a notebook or similiar.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    val analyzerContextWithSuccessfulValues = AnalyzerContext(successfulMetrics)
+
+    AnalysisResultSerde.serialize(
+      Seq(AnalysisResult(Some(resultKey).get, analyzerContextWithSuccessfulValues))
+    )
+  }
+
+  def aggregateMultipleMetricsRepositories(datasets: Seq[DataFrame],
+                                           analyzers: Seq[Analyzer[_, Metric[_]]],
+                                           name: String = "multimetricsrepo.json"): Any = {
```
Comment (velramiir) [inline src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala]: Why does this function have to return _Any_? I think we should avoid using _Any._

File: src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    val objectOne: Seq[AnalysisResult] = AnalysisResultSerde.deserialize(jsonOne)
+    val objectTwo: Seq[AnalysisResult] = AnalysisResultSerde.deserialize(jsonTwo)
+
+    val unioned = objectTwo ++ objectOne
+
+    AnalysisResultSerde.serialize(unioned)
+  }
+
+  def saveRepositoryAsJSON(data: DataFrame, analyzers: Seq[Analyzer[_, Metric[_]]]): String = {
```
Comment (velramiir) [inline src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala]: This method doesn't only save the repository, but also executes an analysis run. I think this should rather be seperate methods

File: src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import com.amazon.deequ.analyzers.runners.{AnalysisRunner, AnalysisRunnerRepositoryOptions, AnalyzerContext}
+import com.amazon.deequ.metrics.Metric
+import com.amazon.deequ.repository.fs.FileSystemMetricsRepository
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.apache.spark.storage.StorageLevel
+
+class MetricsRepositoryDiff (session: SparkSession,
+                             path: String,
+                             saveAllRepositories: Boolean = false) {
+
```
Comment (velramiir) [inline src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala:31]: I'm not quite sure what this conceptionally is. Is it a special kind of metrics repository? Or a class holding multiple repos internally? Perhaps you can make this more clear. If the first is true, I'd suggest to extend the MetricsRepository-Trait.

File: src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    datasets.foreach { data =>
+      val tempJson = saveRepositoryAsJSON(data, analyzers)
+      if (json == "") {
+        json = tempJson
+      }
+      else {
+        json = jsonUnion(json, tempJson)
+      }
+    }
```
Comment (velramiir) [inline src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala:94]: Can you implement the union in a way so this conditional is not necessary? So the union works even if one of the inputs is the empty string?

**Review by velramiir — CHANGES_REQUESTED** [review]

File: src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala (hunk 1 above)
Comment (malte-b) [reply src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala]: You're right, as it doesn't return anything I changed it to _Unit_.

File: src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala (hunk 2 above)
Comment (malte-b) [reply src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala]: I changed the name of the function to better reflect its behaviour. Furthermore I don't think it makes much sense to split it in separate  functions, as the analysis run itself is just executing another function from the AnalysisRunner.

File: src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala (hunk 3 above)
Comment (malte-b) [reply src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala:31]: It is in fact not a special kind of MetricsRepository but rather more of a convenience/helper class, that helps to merge and save multiple MetricsRepositories into one. Therefore it shouldn't extend the MetricsRepository-trait. I get that the naming may seem confusing but we couldn't find a better name for it for now.

File: src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala (hunk 4 above)
Comment (malte-b) [reply src/main/scala/com/amazon/deequ/repository/MetricsRepositoryDiff.scala:94]: The union uses the  AnalysisResultSerde.deserialize function within deequ. This doesn't support the emtpy string as input and as we didn't want to reimplement the whole function just to add that functionality, we decided to skip it on the very first step. This way we also don't have to run the whole function as we can just the initial json variable to tempJson. If you have any suggestions how to better implement it, feel free to leave some feedback!

**Review by velramiir — APPROVED** [review]

**Pull request merged**